com.unity.multiplayer.mlapi/Runtime/Core/RPCQueue/RPCQueueHistoryFrame.cs

```diff
@@ -0,0 +1,226 @@
+            {
+                //NSS: Outbound we care about both channel and clients
+                CurrentQueueItem.Channel = QueueReader.ReadString().ToString();
+                int NumClients = QueueReader.ReadInt32();
+                if (NumClients > 0)
+                {
+                    ulong[] clientIdArray = new ulong[NumClients];
+                    for (int i = 0; i < NumClients; i++)
+                    {
+                        clientIdArray[i] = QueueReader.ReadUInt64();
+                    }
+
+                    if (CurrentQueueItem.ClientIds == null)
+                    {
+                        CurrentQueueItem.ClientIds = clientIdArray;
+                    }
+                    else
+                    {
+                        CurrentQueueItem.ClientIds = clientIdArray;
+                    }
+                }
+                else
+                {
+                    CurrentQueueItem.ClientIds = new ulong[0];
+                }
+            }
+
+            //Get the stream size
+            CurrentQueueItem.StreamSize = QueueReader.ReadInt64();
+
+            //Inbound and Outbound message streams are handled differently
+            if (queueFrameType == QueueFrameType.Inbound)
+            {
+                //Get our offset
+                long Position = QueueReader.ReadInt64();
+                //Always make sure we are positioned at the start of the stream before we write
+                CurrentQueueItem.ItemStream.Position = 0;
+
+                //Write the entire message to the CurrentQueueItem stream (1 stream is re-used for all incoming RPCs)
+                CurrentQueueItem.StreamWriter.ReadAndWrite(QueueReader, CurrentQueueItem.StreamSize);
+
+                //Reset the position back to the offset so std rpc API can process the message properly
+                //(i.e. minus the already processed header)
+                CurrentQueueItem.ItemStream.Position = Position;
+            }
+            else
+            {
+                //Create a byte array segment for outbound sending
+                CurrentQueueItem.MessageData = QueueReader.CreateArraySegment((int)CurrentQueueItem.StreamSize, (int)QueueStream.Position);
+            }
+
+            return CurrentQueueItem;
+        }
+
+        /// <summary>
+        /// GetNextQueueItem
+        /// Handles getting the next queue item from this frame
+        /// If none are remaining, then it returns a queue item type of NONE
+        /// </summary>
+        /// <returns>FrameQueueItem</returns>
+        public FrameQueueItem GetNextQueueItem()
+        {
+            QueueStream.Position = QueueItemOffsets[QueueItemOffsetIndex];
+            QueueItemOffsetIndex++;
+            if (QueueItemOffsetIndex >= QueueItemOffsets.Count)
+            {
+                CurrentQueueItem.QueueItemType = RPCQueueManager.QueueItemType.None;
+                return CurrentQueueItem;
+            }
+
+            return GetCurrentQueueItem();
```

Comment:
ugh, is this an error condition?

Reply:
Not sure what you are seeing here.
It basically checks to see if it has reached the end of the stream and if so then it returns a queue item who's queue item type is "NONE" (i.e. end of entries).